test/jdk/java/awt/im/4959409/bug4959409.java

```diff
@@ -22,51 +22,199 @@
+        robot.waitForIdle();
+        robot.mouseMove(X, Y);
+        robot.delay(DELAY);
+        robot.waitForIdle();
+
+        robot.mousePress(InputEvent.BUTTON1_DOWN_MASK);
+        robot.mouseRelease(InputEvent.BUTTON1_DOWN_MASK);
+        robot.waitForIdle();
     }
 
-    public void keyPressed(KeyEvent e) {
-        int code = e.getKeyCode();
-        int mods = e.getModifiers();
-        if (code == '1' && mods == KeyEvent.SHIFT_MASK) {
-            label.setText("KEYPRESS received for Shift+1");
-        } else {
-            label.setText(" ");
+    public static void checkSwingTopLevelVisible(javax.swing.JFrame jFrame, CountDownLatch topLevelVisibleLatch) throws InterruptedException, InvocationTargetException {
+        SwingUtilities.invokeAndWait(() -> {
+            if (jFrame.isVisible()) {
+                topLevelVisibleLatch.countDown();
+            }
+        });
+    }
+
+    public static boolean isTopLevelVisible(javax.swing.JFrame jFrame, CountDownLatch topLevelVisibleLatch) throws Exception {
+        checkSwingTopLevelVisible(jFrame, topLevelVisibleLatch);
+        if (topLevelVisibleLatch.getCount() != 0) {
+            int count = 1;
+            while (count <= 5) {
+                TimeUnit.SECONDS.sleep(1);
+                checkSwingTopLevelVisible(jFrame, topLevelVisibleLatch);
+                if (topLevelVisibleLatch.getCount() == 0) {
+                    break;
+                }
+                count++;
+            }
         }
+        return topLevelVisibleLatch.await(TIMEOUT, TimeUnit.SECONDS);
     }
 
-    public void keyTyped(KeyEvent e) {
+    public static void checkJComponentVisible(javax.swing.JComponent jComponent, CountDownLatch componentVisibleLatch) throws InterruptedException, InvocationTargetException {
+        SwingUtilities.invokeAndWait(() -> {
+            if (jComponent.isVisible()) {
+                componentVisibleLatch.countDown();
+            }
+        });
     }
 
-    public void keyReleased(KeyEvent e) {
+    public static boolean isJComponentVisible(javax.swing.JComponent jComponent, CountDownLatch componentVisibleLatch) throws InterruptedException, InvocationTargetException {
+        checkJComponentVisible(jComponent, componentVisibleLatch);
+        if (componentVisibleLatch.getCount() != 0) {
+            int count = 1;
+            while (count <= 5) {
+                TimeUnit.SECONDS.sleep(1);
+                checkJComponentVisible(jComponent, componentVisibleLatch);
+                if (componentVisibleLatch.getCount() == 0) {
+                    break;
+                }
+                count++;
+            }
+        }
+        return componentVisibleLatch.await(TIMEOUT, TimeUnit.SECONDS);
+    }
+
+    public static void main(String[] args) throws Exception {
+        try {
+            createUIAndTest();
+        } finally {
+            if (frame != null) {
+                SwingUtilities.invokeAndWait(frame::dispose);
+            }
```

Comment:
This has inconsistency where `frame != null` is run on main thread without synchronisation. The null check should also be run on the EDT as I originally suggested.

Reply:
Done